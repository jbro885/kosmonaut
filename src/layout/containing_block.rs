use crate::layout::rect::Rect;
use crate::style::values::computed::length::CSSPixelLength;
use crate::style::values::computed::{Direction, WritingMode};

/// A containing block is a pairing of a `Rect` and a writing mode that is associated with this
/// block, which is used in determining flow-relative directions for children of this containing
/// block.
///
/// https://www.w3.org/TR/2011/REC-CSS2-20110607/visudet.html#containing-block-details
/// https://drafts.csswg.org/css-writing-modes-4/#logical-direction-layout
#[derive(Copy, Clone, Debug)]
pub struct ContainingBlock {
    rect: Rect,
    direction: Direction,
    writing_mode: WritingMode,
}

impl ContainingBlock {
    pub fn new(rect: Rect, direction: Direction, writing_mode: WritingMode) -> Self {
        Self {
            rect,
            direction,
            writing_mode,
        }
    }

    pub fn direction(&self) -> Direction {
        self.direction
    }

    /// Gets the logical height of this rectangle (also called the block-size).
    pub fn block_size(&self) -> CSSPixelLength {
        if self.writing_mode.is_horizontal() {
            self.rect.height
        } else {
            self.rect.width
        }
    }

    /// Gets the logical width of this rectangle (also called the inline-size).
    pub fn inline_size(&self) -> CSSPixelLength {
        if self.writing_mode.is_horizontal() {
            self.rect.width
        } else {
            self.rect.height
        }
    }

    pub fn rect(&self) -> &Rect {
        &self.rect
    }

    pub fn writing_mode(&self) -> WritingMode {
        self.writing_mode
    }
}
