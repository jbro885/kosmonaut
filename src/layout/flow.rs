use crate::layout::formatting_context::FormattingContextContent;

pub struct BlockFormattingContext {
    /// The contents of this context -- likely a combination of other nested formatting contexts
    /// and boxes.
    contents: Vec<FormattingContextContent>
}

/// A block container either contains only inline-level boxes participating in an inline formatting
/// context, or contains only block-level boxes participating in a block formatting context.
///
/// ### Formatting context establishment
///
/// A block container that contains only inline-level content establishes a new inline formatting
/// context. The element then also generates a root inline box which wraps all of its inline
/// content.
///
/// A block container establishes a new block formatting context if its parent formatting context is
/// not a block formatting context; otherwise, when participating in a block formatting context
/// itself, it either establishes a new block formatting context for its contents or continues the
/// one in which it participates, as determined by the constraints of other properties (such as
/// overflow or align-content).
///
/// https://www.w3.org/TR/css-display/#block-container
pub struct BlockContainer {

}

/// Types of block-level boxes (boxes that participate in block layout).  Note that "participating
/// in block layout" is NOT the same as participating in a block formatting context.  Block-level
/// boxes could be participating in a flex formatting context, grid formatting context, inline
/// formatting context, etc.  In this context, "block layout" refers to the direction in which
/// layout progresses -- in the block direction (y-axis / logical height-wise in the common case).
///
/// https://drafts.csswg.org/css-display/#block-level
pub enum BlockBoxType {
    /// https://drafts.csswg.org/css-display/#block-container
    BlockContainer,
    /// A block-level box that is also a block container (not all block containers are block-level).
    /// https://drafts.csswg.org/css-display/#block-box
    BlockBox
}
