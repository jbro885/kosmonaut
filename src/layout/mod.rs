pub mod box_tree;
pub mod containing_block;
pub mod dimensions;
pub mod flow;
pub mod formatting_context;
pub mod layout_box;
pub mod rect;
pub mod values;

use crate::cli::DumpLayoutVerbosity;
use crate::dom::tree::NodeData;
use crate::layout::containing_block::ContainingBlock;
use crate::layout::flow::OriginRelativeProgression;
use crate::layout::layout_box::LayoutBox;
use crate::layout::rect::Rect;
use crate::style::values::computed::length::CSSPixelLength;
use crate::style::values::CSSFloat;
use std::io::Write;

/// Given a `window` and a `layout_root_box`, perform a layout with the dimensions of the `window`.
pub fn global_layout(
    layout_root_box: &mut LayoutBox,
    inner_window_width: f32,
    inner_window_height: f32,
    scale_factor: f32,
) {
    let writing_mode = layout_root_box.computed_values().writing_mode;
    let direction = layout_root_box.computed_values().direction;
    layout_root_box.layout(LayoutContext::new(ContainingBlock::new(
        Rect {
            start_x: 0.0,
            start_y: 0.0,
            width: CSSPixelLength::new(inner_window_width / scale_factor),
            height: CSSPixelLength::new(inner_window_height / scale_factor),
        },
        direction,
        writing_mode,
    )));
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum BoxComponent {
    Border,
    Margin,
    Padding,
}

#[derive(Copy, Clone, Debug)]
pub struct LayoutContext {
    containing_block: ContainingBlock,
}

impl LayoutContext {
    pub fn new(containing_block: ContainingBlock) -> Self {
        LayoutContext { containing_block }
    }

    pub fn inline_start_origin_relative_progression(&self) -> OriginRelativeProgression {
        self.containing_block
            .inline_start_origin_relative_progression()
    }

    pub fn block_start_origin_relative_progression(&self) -> OriginRelativeProgression {
        self.containing_block
            .block_start_origin_relative_progression()
    }
}

pub trait Layout {
    fn layout(&mut self, context: LayoutContext);
}

/// Trait describing behavior necessary for dumping the layout tree, used in the `dump-layout`
/// tests and debugging.  Should be implemented by "container"-style entities, such as members
/// of the layout tree, formatting individual struct members via the `DumpLayoutFormat` trait.
pub trait DumpLayout {
    fn dump_layout<W: Write>(
        &self,
        write_to: &mut W,
        indent_spaces: usize,
        verbosity: DumpLayoutVerbosity,
    );
}

/// Trait describing behavior necessary for formatting ones data in preparation for a layout tree
/// dump.
/// TODO: Write a custom derive for this.  A bunch of impls of this are just enums calling
/// `dump_layout_format` on their variants.  This trait will still need to be implemented by hand
/// at the leaves, though (sort of like the Debug trait)
/// https://doc.rust-lang.org/book/ch19-06-macros.html#how-to-write-a-custom-derive-macrok
pub trait DumpLayoutFormat {
    fn dump_layout_format(&self) -> String;
}

impl DumpLayoutFormat for CSSFloat {
    fn dump_layout_format(&self) -> String {
        let px = format!("{:.2}", self);
        let mut px_trimmed = px.trim_end_matches('0');
        px_trimmed = px_trimmed.trim_end_matches('.');
        px_trimmed.to_owned()
    }
}

impl DumpLayoutFormat for CSSPixelLength {
    fn dump_layout_format(&self) -> String {
        self.px().dump_layout_format()
    }
}

impl DumpLayoutFormat for NodeData {
    fn dump_layout_format(&self) -> String {
        match self {
            NodeData::Comment(_) => "COMMENT".to_owned(),
            NodeData::Document(_) => "DOCUMENT".to_owned(),
            NodeData::Doctype(_) => "DOCTYPE".to_owned(),
            NodeData::DocumentFragment => "DOCUMENT_FRAGMENT".to_owned(),
            NodeData::Element(element_data) => element_data.name.local.to_uppercase(),
            NodeData::Text(text) => format!("TEXT \"{}\"", text.clone().take().trim()),
            NodeData::ProcessingInstruction(_) => "PROCESSING_INSTRUCTION".to_owned(),
        }
    }
}
