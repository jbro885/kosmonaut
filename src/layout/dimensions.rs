use crate::layout::flow::FlowSide;
use crate::layout::rect::{EdgeSizes, Rect};
use crate::layout::BoxComponent;
use crate::style::values::computed::length::CSSPixelLength;
use crate::style::values::computed::{Direction, WritingMode};
use crate::style::values::CSSFloat;
use crate::Side;

/// https://www.w3.org/TR/2018/WD-css-box-3-20181218/#box-model
#[derive(Clone, Copy, Debug, Default)]
pub struct Dimensions {
    // Position of the content area relative to the document origin:
    pub content: Rect,

    // Surrounding edges:
    pub padding: EdgeSizes,
    pub border: EdgeSizes,
    pub margin: EdgeSizes,
}

impl Dimensions {
    pub fn border_size(self, side: Side) -> CSSPixelLength {
        match side {
            Side::Bottom => self.border.bottom,
            Side::Left => self.border.left,
            Side::Right => self.border.right,
            Side::Top => self.border.top,
        }
    }

    /// The area covered by the content area plus padding and borders.
    pub fn border_box(self) -> Rect {
        self.padding_box().expanded_by(self.border)
    }

    /// The area covered by the content area plus its padding.
    pub fn padding_box(self) -> Rect {
        self.content.expanded_by(self.padding)
    }

    /// The area covered by the content area plus padding, borders, and margin.
    // TODO: This will need to change when we implement margin collapsing: http://www.w3.org/TR/CSS2/box.html#collapsing-margins
    pub fn margin_box(self) -> Rect {
        self.border_box().expanded_by(self.margin)
    }

    pub fn scale_edges_by(&mut self, scale_factor: f32) {
        self.padding.scale_by(scale_factor);
        self.border.scale_by(scale_factor);
        self.margin.scale_by(scale_factor);
    }

    pub fn set_height(&mut self, val: CSSPixelLength) {
        self.content.height = val;
    }

    pub fn set_width(&mut self, val: CSSPixelLength) {
        self.content.width = val;
    }

    pub fn set_start_x(&mut self, val: CSSFloat) {
        self.content.start_x = val;
    }

    pub fn set_start_y(&mut self, val: CSSFloat) {
        self.content.start_y = val;
    }

    pub fn set_block_start_coord(&mut self, val: CSSFloat, writing_mode: WritingMode) {
        match writing_mode {
            WritingMode::HorizontalTb => self.set_start_y(val),
            WritingMode::VerticalRl
            | WritingMode::SidewaysRl
            | WritingMode::VerticalLr
            | WritingMode::SidewaysLr => self.set_start_x(val),
        }
    }

    /// The inline-directions for `horizontal-tb` are left-right, so set `start_x` for that
    /// `writing-mode`.  The inline-directions of the other `writing-mode`s are bottom-top, so
    /// set `start_y` for them here.
    ///
    /// https://github.com/twilco/kosmonaut/blob/master/src/layout/dimensions.rs
    pub fn set_inline_start_coord(&mut self, val: CSSFloat, writing_mode: WritingMode) {
        match writing_mode {
            WritingMode::HorizontalTb => self.set_start_x(val),
            WritingMode::VerticalRl
            | WritingMode::SidewaysRl
            | WritingMode::VerticalLr
            | WritingMode::SidewaysLr => self.set_start_y(val),
        }
    }

    pub fn get_content_block_size(&self, writing_mode: WritingMode) -> CSSPixelLength {
        self.get_block_size(None, writing_mode)
    }

    pub fn padding_box_block_size(&self, writing_mode: WritingMode) -> CSSPixelLength {
        self.get_block_size(Some(BoxComponent::Padding), writing_mode)
    }

    pub fn border_box_block_size(&self, writing_mode: WritingMode) -> CSSPixelLength {
        self.get_block_size(Some(BoxComponent::Border), writing_mode)
    }

    pub fn margin_box_block_size(&self, writing_mode: WritingMode) -> CSSPixelLength {
        self.get_block_size(Some(BoxComponent::Margin), writing_mode)
    }

    /// Gets the block size of these dimensions, optionally expanded by:
    ///
    ///  1. Padding only
    ///  2. Padding and borders
    ///  3. Padding, borders, and margins
    ///
    /// If `None`, only the block size of the box content will be returned.
    ///
    /// Note that the block size is also referred to as the logical height.
    fn get_block_size(
        &self,
        expanded_by: Option<BoxComponent>,
        writing_mode: WritingMode,
    ) -> CSSPixelLength {
        match writing_mode {
            WritingMode::HorizontalTb => match expanded_by {
                None => self.content.height,
                Some(BoxComponent::Padding) => self.padding_box().height,
                Some(BoxComponent::Border) => self.border_box().height,
                Some(BoxComponent::Margin) => self.margin_box().height,
            },
            WritingMode::VerticalRl
            | WritingMode::SidewaysRl
            | WritingMode::VerticalLr
            | WritingMode::SidewaysLr => match expanded_by {
                None => self.content.width,
                Some(BoxComponent::Padding) => self.padding_box().width,
                Some(BoxComponent::Border) => self.border_box().width,
                Some(BoxComponent::Margin) => self.margin_box().width,
            },
        }
    }

    pub fn set_block_size(&mut self, val: CSSPixelLength, writing_mode: WritingMode) {
        match writing_mode {
            WritingMode::HorizontalTb => self.content.height = val,
            WritingMode::VerticalRl
            | WritingMode::SidewaysRl
            | WritingMode::VerticalLr
            | WritingMode::SidewaysLr => self.content.width = val,
        }
    }

    /// Note that the inline size is also known as the logical width.
    pub fn get_inline_size(&self, writing_mode: WritingMode) -> CSSPixelLength {
        match writing_mode {
            WritingMode::HorizontalTb => self.content.width,
            WritingMode::VerticalRl
            | WritingMode::SidewaysRl
            | WritingMode::VerticalLr
            | WritingMode::SidewaysLr => self.content.height,
        }
    }

    pub fn set_inline_size(&mut self, val: CSSPixelLength, writing_mode: WritingMode) {
        match writing_mode {
            WritingMode::HorizontalTb => self.content.width = val,
            WritingMode::VerticalRl
            | WritingMode::SidewaysRl
            | WritingMode::VerticalLr
            | WritingMode::SidewaysLr => self.content.height = val,
        }
    }

    // Make sure the `get` and `set` tables stay in sync.  There's probably a better way to share
    // the lookup logic between get and set operations, but copy-paste will do for now.
    pub fn get(
        &self,
        side: FlowSide,
        box_component: BoxComponent,
        writing_mode: WritingMode,
        direction: Direction,
    ) -> CSSPixelLength {
        // Maps to this table: https://drafts.csswg.org/css-writing-modes-4/#logical-to-physical
        match (writing_mode, direction, side) {
            (
                WritingMode::VerticalRl | WritingMode::SidewaysRl | WritingMode::VerticalLr,
                Direction::Ltr,
                FlowSide::InlineStart,
            )
            | (
                WritingMode::VerticalRl | WritingMode::SidewaysRl | WritingMode::VerticalLr,
                Direction::Rtl,
                FlowSide::InlineEnd,
            )
            | (WritingMode::SidewaysLr, Direction::Ltr, FlowSide::InlineEnd)
            | (WritingMode::SidewaysLr, Direction::Rtl, FlowSide::InlineStart)
            | (WritingMode::HorizontalTb, _, FlowSide::BlockStart) => match box_component {
                BoxComponent::Border => self.border.top,
                BoxComponent::Margin => self.margin.top,
                BoxComponent::Padding => self.padding.top,
            },
            (
                WritingMode::VerticalRl | WritingMode::SidewaysRl | WritingMode::VerticalLr,
                Direction::Ltr,
                FlowSide::InlineEnd,
            )
            | (
                WritingMode::VerticalRl | WritingMode::SidewaysRl | WritingMode::VerticalLr,
                Direction::Rtl,
                FlowSide::InlineStart,
            )
            | (WritingMode::SidewaysLr, Direction::Ltr, FlowSide::InlineStart)
            | (WritingMode::SidewaysLr, Direction::Rtl, FlowSide::InlineEnd)
            | (WritingMode::HorizontalTb, _, FlowSide::BlockEnd) => match box_component {
                BoxComponent::Border => self.border.bottom,
                BoxComponent::Margin => self.margin.bottom,
                BoxComponent::Padding => self.padding.bottom,
            },
            (WritingMode::VerticalRl | WritingMode::SidewaysRl, _, FlowSide::BlockEnd)
            | (WritingMode::VerticalLr | WritingMode::SidewaysLr, _, FlowSide::BlockStart)
            | (WritingMode::HorizontalTb, Direction::Ltr, FlowSide::InlineStart)
            | (WritingMode::HorizontalTb, Direction::Rtl, FlowSide::InlineEnd) => {
                match box_component {
                    BoxComponent::Border => self.border.left,
                    BoxComponent::Margin => self.margin.left,
                    BoxComponent::Padding => self.padding.left,
                }
            }
            (WritingMode::VerticalRl | WritingMode::SidewaysRl, _, FlowSide::BlockStart)
            | (WritingMode::VerticalLr | WritingMode::SidewaysLr, _, FlowSide::BlockEnd)
            | (WritingMode::HorizontalTb, Direction::Ltr, FlowSide::InlineEnd)
            | (WritingMode::HorizontalTb, Direction::Rtl, FlowSide::InlineStart) => {
                match box_component {
                    BoxComponent::Border => self.border.right,
                    BoxComponent::Margin => self.margin.right,
                    BoxComponent::Padding => self.padding.right,
                }
            }
        }
    }

    #[inline(always)]
    pub fn set_margin(
        &mut self,
        side: FlowSide,
        val: CSSPixelLength,
        writing_mode: WritingMode,
        direction: Direction,
    ) {
        self.set(side, BoxComponent::Margin, val, writing_mode, direction)
    }

    #[inline(always)]
    pub fn set_border(
        &mut self,
        side: FlowSide,
        val: CSSPixelLength,
        writing_mode: WritingMode,
        direction: Direction,
    ) {
        self.set(side, BoxComponent::Border, val, writing_mode, direction)
    }

    #[inline(always)]
    pub fn set_padding(
        &mut self,
        side: FlowSide,
        val: CSSPixelLength,
        writing_mode: WritingMode,
        direction: Direction,
    ) {
        self.set(side, BoxComponent::Padding, val, writing_mode, direction);
    }

    pub fn set(
        &mut self,
        side: FlowSide,
        box_component: BoxComponent,
        val: CSSPixelLength,
        writing_mode: WritingMode,
        direction: Direction,
    ) {
        // Maps to this table: https://drafts.csswg.org/css-writing-modes-4/#logical-to-physical
        match (writing_mode, direction, side) {
            (
                WritingMode::VerticalRl | WritingMode::SidewaysRl | WritingMode::VerticalLr,
                Direction::Ltr,
                FlowSide::InlineStart,
            )
            | (
                WritingMode::VerticalRl | WritingMode::SidewaysRl | WritingMode::VerticalLr,
                Direction::Rtl,
                FlowSide::InlineEnd,
            )
            | (WritingMode::SidewaysLr, Direction::Ltr, FlowSide::InlineEnd)
            | (WritingMode::SidewaysLr, Direction::Rtl, FlowSide::InlineStart)
            | (WritingMode::HorizontalTb, _, FlowSide::BlockStart) => match box_component {
                BoxComponent::Border => self.border.top = val,
                BoxComponent::Margin => self.margin.top = val,
                BoxComponent::Padding => self.padding.top = val,
            },
            (
                WritingMode::VerticalRl | WritingMode::SidewaysRl | WritingMode::VerticalLr,
                Direction::Ltr,
                FlowSide::InlineEnd,
            )
            | (
                WritingMode::VerticalRl | WritingMode::SidewaysRl | WritingMode::VerticalLr,
                Direction::Rtl,
                FlowSide::InlineStart,
            )
            | (WritingMode::SidewaysLr, Direction::Ltr, FlowSide::InlineStart)
            | (WritingMode::SidewaysLr, Direction::Rtl, FlowSide::InlineEnd)
            | (WritingMode::HorizontalTb, _, FlowSide::BlockEnd) => match box_component {
                BoxComponent::Border => self.border.bottom = val,
                BoxComponent::Margin => self.margin.bottom = val,
                BoxComponent::Padding => self.padding.bottom = val,
            },
            (WritingMode::VerticalRl | WritingMode::SidewaysRl, _, FlowSide::BlockEnd)
            | (WritingMode::VerticalLr | WritingMode::SidewaysLr, _, FlowSide::BlockStart)
            | (WritingMode::HorizontalTb, Direction::Ltr, FlowSide::InlineStart)
            | (WritingMode::HorizontalTb, Direction::Rtl, FlowSide::InlineEnd) => {
                match box_component {
                    BoxComponent::Border => self.border.left = val,
                    BoxComponent::Margin => self.margin.left = val,
                    BoxComponent::Padding => self.padding.left = val,
                }
            }
            (WritingMode::VerticalRl | WritingMode::SidewaysRl, _, FlowSide::BlockStart)
            | (WritingMode::VerticalLr | WritingMode::SidewaysLr, _, FlowSide::BlockEnd)
            | (WritingMode::HorizontalTb, Direction::Ltr, FlowSide::InlineEnd)
            | (WritingMode::HorizontalTb, Direction::Rtl, FlowSide::InlineStart) => {
                match box_component {
                    BoxComponent::Border => self.border.right = val,
                    BoxComponent::Margin => self.margin.right = val,
                    BoxComponent::Padding => self.padding.right = val,
                }
            }
        }
    }
}
