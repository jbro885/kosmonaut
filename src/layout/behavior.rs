use crate::dom::tree::NodeRef;
use crate::layout::containing_block::ContainingBlock;
use crate::layout::dimensions::Dimensions;
use crate::layout::formatting_context::FormattingContextRef;
use crate::style::values::computed::ComputedValues;
use accountable_refcell::Ref;
use enum_dispatch::enum_dispatch;

#[enum_dispatch(LayoutBox, BlockLevelBox, InlineLevelBox)]
pub trait ApplyPageRelativeProperties {
    fn apply_block_page_relative_properties(&mut self, containing_block: ContainingBlock);
    fn apply_inline_page_relative_properties(&mut self, containing_block: ContainingBlock);
}

#[macro_export]
macro_rules! apply_page_relative_properties_base_box_passthrough_impls {
    () => {
        #[inline(always)]
        fn apply_block_page_relative_properties(&mut self, containing_block: ContainingBlock) {
            self.base
                .apply_block_page_relative_properties(containing_block);
        }

        #[inline(always)]
        fn apply_inline_page_relative_properties(&mut self, containing_block: ContainingBlock) {
            self.base
                .apply_inline_page_relative_properties(containing_block);
        }
    };
}

/// Behaviors applicable to all members of the box-tree (including text runs).  Because this
/// encompasses a wide variation of box types, some basic behaviors, like anything to do with
/// children, are not present in this trait.  This is because some boxes may have children of a more
/// specific type than `LayoutBox` (e.g. only `BlockLevelBox`s or `InlineLevelBox`s), or have no
/// children at all (e.g. for text runs).
#[enum_dispatch(LayoutBox, BlockLevelBox, InlineLevelBox, InlineLevelContent)]
pub trait BaseLayoutBoxBehavior {
    fn computed_values(&self) -> Ref<ComputedValues>;
    fn dimensions(&self) -> Dimensions;
    fn dimensions_mut(&mut self) -> &mut Dimensions;
    fn formatting_context(&self) -> FormattingContextRef;
    fn is_root(&self) -> bool;
    fn node(&self) -> NodeRef;
}

#[macro_export]
macro_rules! layout_box_behavior_base_box_passthrough_impls {
    () => {
        #[inline(always)]
        fn computed_values(&self) -> Ref<ComputedValues> {
            self.base.computed_values()
        }

        #[inline(always)]
        fn dimensions(&self) -> Dimensions {
            self.base.dimensions()
        }

        #[inline(always)]
        fn dimensions_mut(&mut self) -> &mut Dimensions {
            self.base.dimensions_mut()
        }

        #[inline(always)]
        fn formatting_context(&self) -> FormattingContextRef {
            self.base.formatting_context()
        }

        #[inline(always)]
        fn is_root(&self) -> bool {
            self.base.is_root()
        }

        #[inline(always)]
        fn node(&self) -> NodeRef {
            self.base.node()
        }
    };
}
