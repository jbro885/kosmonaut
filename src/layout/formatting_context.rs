use crate::layout::dimensions::{LogicalDimensions, PhysicalDimensions};
use crate::layout::layout_box::LayoutBox;
use image::error::UnsupportedErrorKind::Format;
use std::mem::discriminant;

/// A formatting context is the environment into which a set of related boxes are laid out.
/// Different formatting contexts lay out their boxes according to different rules.
///
/// https://drafts.csswg.org/css-display/#formatting-context
#[derive(Copy, Clone, Debug, PartialEq)]
pub enum FormattingContext {
    Block,
    Inline,
}

/// A formatting context can contain sub-formatting-contexts and boxes.
pub enum FormattingContextContent {
    SubContext(QualifiedFormattingContext),
    /// The root of some amount of boxes.
    Box(LayoutBox),
}

#[derive(Copy, Clone, Debug, PartialEq)]
pub enum QualifiedFormattingContext {
    /// Layout of this formatting context is independent of its parent context.
    /// https://www.w3.org/TR/css-display-3/#independent-formatting-context
    Independent(FormattingContext),
    /// Layout of this formatting context is dependent on the formatting context of its parent.  For
    /// example, inline formatting contexts are part of (and therefore "dependent" on) their parent
    /// block formatting context.
    /// https://www.w3.org/TR/css-display-3/#inline-formatting-context
    Dependent(FormattingContext),
}

impl QualifiedFormattingContext {
    pub fn is_inline_formatting_context(&self) -> bool {
        match self {
            QualifiedFormattingContext::Independent(fc)
            | QualifiedFormattingContext::Dependent(fc) => fc == &FormattingContext::Inline,
        }
    }
}
