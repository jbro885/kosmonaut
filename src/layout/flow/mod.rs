pub mod block;
pub mod inline;

use crate::dom::tree::NodeRef;
use crate::layout::containing_block::ContainingBlock;
use crate::layout::formatting_context::FormattingContextRef;
use crate::layout::layout_box::{
    AnonymousBlockBox, AnonymousInlineBox, BaseBox, BlockContainer, InlineBox, LayoutBox, TextRun,
};
use crate::style::values::computed::length::{
    CSSPixelLength, LengthPercentage, LengthPercentageOrAuto,
};
use crate::style::values::computed::Direction;
use crate::style::values::used::ToPx;
use crate::base_box_passthrough_impls;

/// A box that contains either contains only inline-level boxes participating in an inline
/// formatting context, or contains only block-level boxes participating in a block formatting
/// context (possibly generating anonymous block boxes to ensure this constraint, as defined in
/// [CSS2ยง9.2.1.1](https://www.w3.org/TR/CSS2/visuren.html#anonymous-block-level).
///
/// https://drafts.csswg.org/css-display-3/#block-container
#[derive(Clone, Debug)]
pub struct BlockContainer {
    base: BaseBox,
    children: Vec<LayoutBox>,
}

impl BlockContainer {
    base_box_passthrough_impls!();

    pub fn new(node: NodeRef, fc: FormattingContextRef) -> Self {
        BlockContainer {
            base: BaseBox::new(node, fc),
            children: Vec::new(),
        }
    }
    
    pub fn add_child(&mut self, new_child: LayoutBox) {
        self.children.push(new_child)
    }

    pub fn children(&self) -> &Vec<LayoutBox> {
        &self.children
    }
}

/// Sides relative to the flow of the page, rather than physical sides (e.g. left, top, ...).
///
/// https://drafts.csswg.org/css-writing-modes-4/#logical-directions
#[derive(Copy, Clone, Debug, PartialEq)]
pub enum FlowSide {
    BlockStart,
    BlockEnd,
    InlineStart,
    InlineEnd,
}
