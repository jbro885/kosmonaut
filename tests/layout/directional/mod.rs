use crate::layout::{dump_layout_cmd, DumpLayoutVerbosity};
use insta::assert_snapshot;

const DIRECTIONAL_WEBSRC_DIR: &str = "tests/websrc/directional";
fn path(filename: &str) -> String {
    format!("{}/{}", DIRECTIONAL_WEBSRC_DIR, filename)
}

#[test]
fn ltr_vertical_lr_block_boxes() {
    let mut dump_layout_cmd = dump_layout_cmd(DumpLayoutVerbosity::NonVerbose);
    dump_layout_cmd
        .arg(path("ltr-vertical-lr-block-boxes.html"))
        .succeeds();
    assert_snapshot!(dump_layout_cmd.stdout());
}

#[test]
fn ltr_vertical_lr_block_boxes_bottom_right_mbp_applied_physically() {
    let mut dump_layout_cmd = dump_layout_cmd(DumpLayoutVerbosity::Verbose);
    dump_layout_cmd
        .arg(path("ltr-vertical-lr-block-boxes-bottom-right-mbp-applied-physically.html"))
        .succeeds();
    assert_snapshot!(dump_layout_cmd.stdout());
}

#[test]
fn ltr_vertical_lr_block_boxes_top_left_right_mbp_applied_physically() {
    let mut dump_layout_cmd = dump_layout_cmd(DumpLayoutVerbosity::Verbose);
    dump_layout_cmd
        .arg(path("ltr-vertical-lr-block-boxes-top-left-mbp-applied-physically.html"))
        .succeeds();
    assert_snapshot!(dump_layout_cmd.stdout());
}

#[test]
fn rtl_horizontal_tb_block_boxes() {
    let mut dump_layout_cmd = dump_layout_cmd(DumpLayoutVerbosity::NonVerbose);
    dump_layout_cmd
        .arg(path("rtl-horizontal-tb-block-boxes.html"))
        .succeeds();
    assert_snapshot!(dump_layout_cmd.stdout());
}
