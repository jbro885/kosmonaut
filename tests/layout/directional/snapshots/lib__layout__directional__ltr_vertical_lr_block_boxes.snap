---
source: tests/layout/directional/mod.rs
expression: dump_layout_cmd.stdout()
---
HTML BlockContainer at (0, 0) size 1006x1080
  BODY BlockContainer at (8, 8) size 990x1064
    DIV BlockContainer at (23, 23) size 300x200
    DIV BlockContainer at (353, 23) size 300x200
    DIV BlockContainer at (683, 23) size 300x200

